File: typescript/4-tidy-first/src/event-ticket-service.ts
// NotificationService handles sending emails
export class NotificationService {
  // the connected flag
  private connected: boolean = false;
  // stores the host
  private host: string = "";
  // list of sent messages
  public sentMessages: string[] = [];

  // connects to the mail server
  connect(host: string, port: number, useTls: boolean): void {
    this.host = host;
    this.connected = true;
  }

  // sends an email
  sendMail(to: string, subject: string, body: string): void {
    if (this.connected) {
      this.sentMessages.push(`To: ${to} | Subject: ${subject} | Body: ${body}`);
    }
  }

  // disconnects from the mail server
  disconnect(): void {
    this.connected = false;
    this.host = "";
  }
}

// EventTicketService manages event tickets
export class EventTicketService {
  // counter for events
  private eventCounter: number = 0;
  // the notification service
  private notificationService: NotificationService;

  // generates a confirmation message
  generateConfirmationMessage(eventName: string, ticketCount: number, totalPrice: number): string {
    return `Booking confirmed: ${ticketCount} ticket(s) for "${eventName}" – Total: €${totalPrice.toFixed(2)}`;
  }

  // calculates the ticket price
  calculateTicketPrice(options: Record<string, any>): number {
    // initialize finalPrice to 0
    let finalPrice = 0;
    // check if quantity is greater than 0
    if (options["quantity"] > 0) {
      finalPrice = options["basePrice"] * options["quantity"] * (options["isEarlyBird"] ? (1 - 0.15) : 1) * (options["isVip"] ? 1.5 : 1) * (options["quantity"] >= 10 ? 0.85 : 1) * 1.19;
    }
    // round to 2 decimal places
    return Math.round(finalPrice * 100) / 100;
  }

  // generates a summary message for the event
  generateSummaryMessage(eventName: string, date: string, venue: string): string {
    // create an array of parts
    const parts: string[] = [];
    parts.push("Event Summary:");
    parts.push(eventName);
    parts.push("Date: " + date);
    parts.push("Venue: " + venue);
    // join with pipe separator
    return parts.join(" | ");
  }

  // sends a booking confirmation email
  sendBookingConfirmation(email: string, eventName: string, ticketCount: number, totalPrice: number): void {
    // connect to smtp server
    this.notificationService.connect("smtp.tickets.example.com", 587, true);
    // send the email
    this.notificationService.sendMail(
      email,
      "Booking Confirmation – " + eventName,
      this.generateConfirmationMessage(eventName, ticketCount, totalPrice)
    );
    // disconnect
    this.notificationService.disconnect();
  }

  // formats a date as dd.mm.yyyy
  formatEventDate(date: Date): string {
    // get the day
    const day = date.getDate().toString().padStart(2, "0");
    // get the month
    const month = (date.getMonth() + 1).toString().padStart(2, "0");
    // get the year
    const year = date.getFullYear();
    return `${day}.${month}.${year}`;
  }

  // constructor
  constructor(notificationService: NotificationService) {
    this.notificationService = notificationService;
  }

  // generates a receipt message
  generateReceiptMessage(eventName: string, ticketCount: number, totalPrice: number, customerName: string): string {
    return "Receipt for " + customerName + " – " + ticketCount.toString() + "x " + eventName + " – Total: €" + totalPrice.toFixed(2);
  }
}

File: python/4-tidy-first/src/event_ticket_service.py
class NotificationService:
    def __init__(self):
        self.sent_messages = []
        self._connected = False
        self._host = ""

    def connect(self, host, port, use_tls):
        self._host = host
        self._connected = True

    def send_mail(self, to, subject, body):
        if self._connected:
            self.sent_messages.append(f"To: {to} | Subject: {subject} | Body: {body}")

    def disconnect(self):
        self._connected = False
        self._host = ""


class EventTicketService:
    event_counter = 0

    def generate_confirmation_message(self, event_name, ticket_count, total_price):
        return f'Booking confirmed: {ticket_count} ticket(s) for "{event_name}" – Total: €{total_price:.2f}'

    def calculate_ticket_price(self, options):
        final_price = 0
        if options["quantity"] > 0:
            early_bird_factor = (1 - 0.15) if options.get("is_early_bird") else 1
            vip_factor = 1.5 if options.get("is_vip") else 1
            group_factor = 0.85 if options["quantity"] >= 10 else 1
            final_price = options["base_price"] * options["quantity"] * early_bird_factor * vip_factor * group_factor * 1.19
        return round(final_price, 2)

    def generate_summary_message(self, event_name, date, venue):
        parts = []
        parts.append("Event Summary:")
        parts.append(event_name)
        parts.append("Date: " + date)
        parts.append("Venue: " + venue)
        return " | ".join(parts)

    def send_booking_confirmation(self, email, event_name, ticket_count, total_price):
        self.notification_service.connect("smtp.tickets.example.com", 587, True)
        self.notification_service.send_mail(
            email,
            "Booking Confirmation – " + event_name,
            self.generate_confirmation_message(event_name, ticket_count, total_price)
        )
        self.notification_service.disconnect()

    def format_event_date(self, day, month, year):
        return f"{day:02d}.{month:02d}.{year}"

    def __init__(self, notification_service):
        self.notification_service = notification_service

    def generate_receipt_message(self, event_name, ticket_count, total_price, customer_name):
        return "Receipt for " + customer_name + " – " + str(ticket_count) + "x " + event_name + " – Total: €" + f"{total_price:.2f}"
